### PR TITLE
Use `github.com/tinkerbell/dhcp` library:

### PR DESCRIPTION
**Pull request by jacobweinstock**
## Description

This is an improved DHCP library. The package layout has been updated to improve the reflection of the services that Boots runs.

`test` directory and docker-compose were updated to work with these changes.

The cli flags and env vars have been updated and reorganized by function (syslog, dhcp, tftp, http ipxe binaries, http ipxe scripts). This helps func main be to more understandable.

The phonehome endpoint and all `/_packet` were removed. These were old deprecated endponts that don't serve much purpose now.

The syslog package was updated to handle shutdown via context, so that it can be used like the other services with `errgroup`.

Update nix shell file to use Go 1.20.

## Why is this needed

Fixes: #266
Fixes: #195

## Checklist:

I have:

- [ ] updated the documentation and/or roadmap (if required)
- [ ] added unit or e2e tests
- [ ] provided instructions on how to upgrade

### CODE REVIEW

**Comment by jacobweinstock** [comment]
Hey @chrisdoherty4, apologies that this PR is so large. It's a significant library change so i was unable to avoid this churn. We can go over it together if that would help. I can do this in our open forum community meeting if that works.

File: .golangci.yml [hunk 1]
```diff
@@ -106,7 +106,8 @@ linters:
     - asciicheck
     - bodyclose
     - cyclop
-    - deadcode
+    # deprecated
+    # - deadcode
```
Comment (chrisdoherty4) [inline .golangci.yml:110]: Do you want to just delete these? I don't think there's a reason to leave them in given we don't intend to reenable.

File: cmd/boots/flag_test.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  -dhcp-syslog-ip             [dhcp] syslog server IP address to use in DHCP packets (opt 7) (default "%[1]v")
+  -dhcp-tftp-ip               [dhcp] tftp server IP address to use in DHCP packets (opt 66, etc) (default "%[1]v:69")
+  -extra-kernel-args          [http] extra set of kernel args (k=v k=v) that are appended to the kernel cmdline iPXE script
+  -http-addr                  [http] local IP and port to listen on for iPXE http script requests (default "%[1]v:80")
+  -http-ipxe-binary-enabled   [http] enable iPXE http binary server (default "true")
+  -http-ipxe-script-enabled   [http] enable iPXE http script server) (default "true")
+  -osie-url                   [http] url where OSIE(Hook) images are located
+  -tink-server                [http] ip:port for the Tink server
+  -tink-server-tls            [http] use TLS for Tink server (default "false")
+  -trusted-proxies            [http] comma separated list of trusted proxies
```
Comment (chrisdoherty4) [inline cmd/boots/flag_test.go:101]: IPs or CIDRs or Domains?

File: cmd/boots/flag_test.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  -dhcp-enabled               [dhcp] enable DHCP server (default "true")
+  -dhcp-http-ipxe-binary-ip   [dhcp] http ipxe binary server IP address to use in DHCP packets (default "http://%[1]v:8080/ipxe/")
+  -dhcp-http-ipxe-script-url  [dhcp] http ipxe script server URL to use in DHCP packets (default "http://%[1]v/auto.ipxe")
+  -dhcp-ip-for-packet         [dhcp] ip address to use in DHCP packets (opt 54, etc) (default "%[1]v")
+  -dhcp-syslog-ip             [dhcp] syslog server IP address to use in DHCP packets (opt 7) (default "%[1]v")
+  -dhcp-tftp-ip               [dhcp] tftp server IP address to use in DHCP packets (opt 66, etc) (default "%[1]v:69")
+  -extra-kernel-args          [http] extra set of kernel args (k=v k=v) that are appended to the kernel cmdline iPXE script
+  -http-addr                  [http] local IP and port to listen on for iPXE http script requests (default "%[1]v:80")
+  -http-ipxe-binary-enabled   [http] enable iPXE http binary server (default "true")
+  -http-ipxe-script-enabled   [http] enable iPXE http script server) (default "true")
```
Comment (chrisdoherty4) [inline cmd/boots/flag_test.go:97]:
```suggestion
  -http-ipxe-script-enabled   [http] enable iPXE http script server (default "true")
```

File: cmd/boots/flag_test.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+  -osie-url                   [http] url where OSIE(Hook) images are located
```
Comment (chrisdoherty4) [inline cmd/boots/flag_test.go:98]:
```suggestion
  -osie-url                   [http] URL where OSIE(Hook) images are located
```

File: cmd/boots/flag_test.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+		t.Fatal(diff)
+	}
+}
+
+func TestCustomUsageFunc(t *testing.T) {
+	defaultIP := detectPublicIPv4("")
+	want := fmt.Sprintf(`USAGE
+  Run Boots server for provisioning
+
+FLAGS
```
Comment (chrisdoherty4) [inline cmd/boots/flag_test.go:79]: There are a number of acronym inconsistencies in the help. `ip` vs `IP`, `http` vs `HTTP`.

File: cmd/boots/main.go [hunk 6]
```diff
@@ -104,173 +107,209 @@ func main() {
...
+	}
 
-	default:
-		return nil, fmt.Errorf("must specify DATA_MODEL_VERSION")
+	return &dhcp.Listener{Addr: bindAddr}, dh, nil
+}
+
+// ipxeScript is is needed to be able to translate the handler.BackendReader
+// returned data to the script.Data struct.
+type ipxeScript struct {
```
Comment (chrisdoherty4) [inline cmd/boots/main.go:289]: It seems odd for this to live here?

**Review by chrisdoherty4 — APPROVED** [review]
Minimal feedback as everything seems good.

File: cmd/boots/main.go (hunk 6 above)
Comment (jacobweinstock) [reply cmd/boots/main.go]: ah yes, updated to a better place.

File: cmd/boots/flag_test.go (hunk 5 above)
Comment (jacobweinstock) [reply cmd/boots/flag_test.go:79]: fixed

File: cmd/boots/flag_test.go (hunk 4 above)
Comment (jacobweinstock) [reply cmd/boots/flag_test.go]: fixed

File: cmd/boots/flag_test.go (hunk 3 above)
Comment (jacobweinstock) [reply cmd/boots/flag_test.go]: fixed

File: cmd/boots/flag_test.go (hunk 2 above)
Comment (jacobweinstock) [reply cmd/boots/flag_test.go]: CIDRs. added to the description.

File: cmd/boots/flag.go [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+	fs.StringVar(&c.dhcp.bindAddr, "dhcp-addr", "0.0.0.0:67", "[dhcp] local IP and port to listen on for DHCP requests")
+	fs.StringVar(&c.dhcp.ipForPacket, "dhcp-ip-for-packet", detectPublicIPv4(""), "[dhcp] ip address to use in DHCP packets (opt 54, etc)")
+	fs.StringVar(&c.dhcp.syslogIP, "dhcp-syslog-ip", detectPublicIPv4(""), "[dhcp] syslog server IP address to use in DHCP packets (opt 7)")
+	fs.StringVar(&c.dhcp.tftpIP, "dhcp-tftp-ip", detectPublicIPv4(":69"), "[dhcp] tftp server IP address to use in DHCP packets (opt 66, etc)")
+	fs.StringVar(&c.dhcp.httpIpxeBinaryIP, "dhcp-http-ipxe-binary-ip", "http://"+detectPublicIPv4(":8080/ipxe/"), "[dhcp] http ipxe binary server IP address to use in DHCP packets")
+	fs.StringVar(&c.dhcp.httpIpxeScriptURL, "dhcp-http-ipxe-script-url", "http://"+detectPublicIPv4("/auto.ipxe"), "[dhcp] http ipxe script server URL to use in DHCP packets")
+}
+
+func backendFlags(c *config, fs *flag.FlagSet) {
+	fs.BoolVar(&c.backends.file.Enabled, "backend-file-enabled", false, "[backend] enable the file backend for DHCP and the HTTP iPXE script")
```
[the remark this answers is not in the document]
Comment (jacobweinstock) [reply cmd/boots/flag.go:118]: yes, correct. fixed now.

File: .golangci.yml (hunk 1 above)
Comment (jacobweinstock) [reply .golangci.yml]: ah yes, fixed.

**Review by chrisdoherty4 — APPROVED** [review]

**Pull request merged**